### PR TITLE
Repaired NBNS and some minor changes

### PR DESCRIPTION
**Pull request by htibosch**
PR Discontinued
---

This PR will be discontinued because it has too many changes, which makes it difficult to merge it. I will split it up and present it in smaller parts.

I already created [PR #817](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/pull/817), which is about DNS and NBNS.
The next PR will be about DHCPv6.

Description
-----------
● FreeRTOS_DNS_Parser.c
Repaired the NBNS functionality.

● FreeRTOS_IP.c
Added some asserts and removed unnecessary brackets.

● FreeRTOS_DNS.c
Changed:
~~~diff
-    const char * pcDot = ( const char * ) strchr( pcHostName, ( int32_t ) '.' );
+    const char * pcDot = ( const char * ) strchr( pcHostName, ( int ) '.' );
~~~

Because `strchr()` expects an int. Checked with Coverity.

~~~diff
-    DNSBuffer_t xReceiveBuffer = { 0 };
+    DNSBuffer_t xReceiveBuffer;
+    ( void ) memset( ( void * ) &xReceiveBuffer, 0, sizeof( xReceiveBuffer ) );
~~~
Because sometimes the initialisation of a struct using `{ 0 }` may lead to an exception because of unaligned access.

● FreeRTOS_ARP.c
Removed unnecessary casts and brackets. Checked with Coverity.

● FreeRTOS_DHCP.c
Use `xSocketValid()` wherever possible.

● FreeRTOS_IPv6_Sockets.c
Do not access IPv4 field `xIPAddress.ulIP_IPv4`
Assume that the packet is proven to have the type `ipIPv6_FRAME_TYPE`.
Change because of MISRA rule:
~~~diff
-    if( xSocketValid( pxSocket ) )
+    if( xSocketValid( pxSocket ) == pdTRUE )
~~~

● FreeRTOS_DNS_Cache.c
Function `prvReadDNSCache()`: chang logic of building up a linked list of addresses.
Assert that `ppxAddressInfo != NULL`.

● FreeRTOS_TCP_Transmission_IPV6.c
Check if pxIPHeader is no NULL before changing it

● FreeRTOS_UDP_IPv6.c
In `vProcessGeneratedUDPPacket_IPv6()`, consider using `pxNetworkBuffer->pxEndPoint`

● FreeRTOS_IP_Utils.c
removed unnecessary ```#if defined```:

~~~diff
-    #if defined( ipconfigIPv4_BACKWARD_COMPATIBLE ) && ( ipconfigIPv4_BACKWARD_COMPATIBLE == 1 )
+    #if ( ipconfigIPv4_BACKWARD_COMPATIBLE == 1 )
~~~

● FreeRTOS_TCP_Transmission_IPV4.c
    Check if pxIPHeader != NULL before changing it

● FreeRTOS_TCP_Utils_IPV6.c
    Make sure MSS doesn't get negative

● FreeRTOS_DNS_Callback.c
    Removed unnecessary brackets
    
● FreeRTOS_IPv4.c
    Removed an unnecessary cast to "( const uint8_t * )".

● FreeRTOS_TCP_State_Handling_IPV4.c
● FreeRTOS_TCP_State_Handling_IPV6.c

Avoid the use of `FREERTOS_INVALID_SOCKET`:
~~~diff
-    if( ( pxNewSocket == NULL ) || ( pxNewSocket == FREERTOS_INVALID_SOCKET ) )
+    if( xSocketValid( pxNewSocket ) == pdFALSE )
~~~

● BufferAllocation_2.c
Some changes that may help clarity and safety in the function `vReleaseNetworkBufferAndDescriptor()`

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by htibosch** [comment]
About the commit [Repaired DHCPv6](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/pull/800/commits/c6e7de47482a89af30488a08edb1631c26125694):

- source/FreeRTOS_ARP.c : xCheckRequiresARPResolution()
    vNDSendNeighbourSolicitation got called with the wrong network buffer
~~~
-    vNDSendNeighbourSolicitation( pxNetworkBuffer, pxIPAddress );
+    vNDSendNeighbourSolicitation( pxTempBuffer, pxIPAddress );
~~~

- source/FreeRTOS_DHCP.c :  adapted `xIsDHCPSocket()` to work for both IPv4 and IPv6
~~~
    /* 'usLocalPort' is host-endian. */
    uint16_t usPortNr = xSocket->usLocalPort;

    /* DHCP clients have the UDP port number 68 */
    if( ( usPortNr == ipDHCP_CLIENT_PORT ) ||
        ( usPortNr == ipDHCPv6_CLIENT_PORT ) )
    {
    }
~~~

- source/include/FreeRTOS_DNS_Globals.h : renamed macros:

    `ipDHCP_CLIENT` -> `ipDHCP_CLIENT_PORT`
    `ipDHCP_SERVER` -> `ipDHCP_SERVER_PORT`

Added 2 macros:
~~~
    /** @brief Default v6 DHCP client port. */
    #define ipDHCPv6_CLIENT_PORT     546U
    /** @brief Default v6 DHCP server port. */
    #define ipDHCPv6_SERVER_PORT     547U
~~~

- source/include/FreeRTOS_DHCPv6.h : make sure that multiple DNS addresses can be received.
  At most `ipconfigENDPOINT_DNS_ADDRESS_COUNT` (existing macro) can be stored.

- source/FreeRTOS_DHCPv6.c :
    Check the result of `xBitConfig_init()`
    Do not assign a struct to a struct, use memcpy in stead.
~~~diff
-    *( pxIterator->pxDHCPMessage ) = xDHCPMessage;
+    memcpy( pxIterator->pxDHCPMessage, &xDHCPMessage, sizeof xDHCPMessage );
~~~

- source/FreeRTOS_DHCPv6.c :
    In `vDHCPv6ProcessEndPoint_HandleReply()`, copy the DNS addresses to the endpoint.
    In state eWaitingOffer, make sure that the timeout is well checked,
        and make sure that the solicit request is repeated.
    When multiple DNS addresses are received, they will all be stored,
    up to `ipconfigENDPOINT_DNS_ADDRESS_COUNT` entries.
    
- source/FreeRTOS_ND.c :
   pxFindLocalEndpoint() now returns the first IPv6 endpoint of the type `eIPv6_LinkLocal`
   Neighbour solicitations for "fe80::1" with a matching MAC-address should be replied to.

- source/FreeRTOS_UDP_IPv6.c :
    When selecting a matching endpoint, also look at global versus non-global addresses.
    `vNDSendNeighbourSolicitation()` (line 344) should not be called because
        `prvStartLookup()` has already sent the "neighbor solicitation".
    `xProcessReceivedUDPPacket_IPv6()` : cleaned up the code a little.
    `uxIPHeaderSizePacket()` should not be called from an IPv6 module.

- source/include/FreeRTOS_IPv6_Sockets.h :
    Added declaration of FreeRTOS_inet_pton6(), just before the existing `FreeRTOS_inet_ntop6()`.

**Review by ActoryOu — CHANGES_REQUESTED** [review]

File: source/FreeRTOS_DNS_Parser.c [hunk 1]
```diff
@@ -1112,105 +1135,115 @@
...
-                            }
-                        }
-                    }
+                /* NBNS only handles IPv4 or "A" records. */
+                xEndPoint.bits.bIPv6 = pdFALSE_UNSIGNED;
+                xEndPoint.usDNSType = dnsTYPE_A_HOST;
 
-                    /* Should not occur: pucUDPPayloadBuffer is part of a xNetworkBufferDescriptor */
-                    if( pxNetworkBuffer != NULL )
+                #if defined( ipconfigIPv4_BACKWARD_COMPATIBLE ) && ( ipconfigIPv4_BACKWARD_COMPATIBLE == 1 )
```
[the remark this answers is not in the document]
Comment (htibosch) [reply source/FreeRTOS_DNS_Parser.c:1172]: As I commented several times, all ipconfig macros **must** be defined, either in the user's FreeRTOSIPConfig.h, or in FreeRTOSIPConfigDefaults.h.
Therefore it is not necessary to test `#if defined` for a ipconfig macro.

I know that many other projects use `#ifdef` and do not test the value of the configuration macros. Not so in FreeRTOS. A zero means disabled, and non-zero means enabled.

**Review by htibosch — COMMENTED** [review]
Thank you Actory for this useful review.

File: source/FreeRTOS_DNS_Parser.c [hunk 2]
```diff
@@ -488,13 +496,13 @@
...
                          * request. Ask the application if it uses the name. */
-                        #if defined( ipconfigIPv4_BACKWARD_COMPATIBLE ) && ( ipconfigIPv4_BACKWARD_COMPATIBLE == 1 )
-                            lDNSHookReturn = xApplicationDNSQueryHook( xSet.pcName );
+                        #if ( ipconfigIPv4_BACKWARD_COMPATIBLE == 1 )
+                            xDNSHookReturn = xApplicationDNSQueryHook( xSet.pcName );
                         #else
-                            lDNSHookReturn = xApplicationDNSQueryHook_Multi( &xEndPoint, xSet.pcName );
-                        #endif /* if defined( ipconfigIPv4_BACKWARD_COMPATIBLE ) && ( ipconfigIPv4_BACKWARD_COMPATIBLE == 1 ) */
+                            xDNSHookReturn = xApplicationDNSQueryHook_Multi( &xEndPoint, xSet.pcName );
+                        #endif
```
[the remark this answers is not in the document]
Comment (ActoryOu) [reply source/FreeRTOS_DNS_Parser.c:503]: It's fine when area of #if to #endif is small. Just a minor suggestion to align coding style.

File: source/FreeRTOS_DNS_Parser.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (ActoryOu) [reply source/FreeRTOS_DNS_Parser.c:1172]: Agree. So the `defined( ipconfigIPv4_BACKWARD_COMPATIBLE ) && ` should be removed, doesn't it?

File: source/FreeRTOS_IP.c [hunk 3]
```diff
@@ -771,7 +771,7 @@ BaseType_t FreeRTOS_NetworkDownFromISR( struct xNetworkInterface * pxNetworkInte
 }
 /*-----------------------------------------------------------*/
 
-#if defined( ipconfigIPv4_BACKWARD_COMPATIBLE ) && ( ipconfigIPv4_BACKWARD_COMPATIBLE == 1 )
+#if ( ipconfigIPv4_BACKWARD_COMPATIBLE == 1 )
```
Comment (shubnil) [inline source/FreeRTOS_IP.c:774]: We want the define option to be present as this will help the FreeRTOS+TCP main branch code to run seamlessly.

**Review by shubnil — CHANGES_REQUESTED** [review]
HI Hein, It will be good if you can create a change only with NBNS changes. This change being big and does a lot of things and will also break the unit tests in almost files.

**Comment by htibosch** [comment]
Closing this PR because it will be committed in multiple smaller PR's.
I will keep my #817 up to date for those who want to play with the demo.

**Pull request closed without merging**